File: nextjs-frontend/lib/shareUtils.ts
/**
 * Utility functions for sharing posts to various social media platforms
 */

export interface ShareOptions {
  url: string;
  title: string;
  text?: string;
  hashtags?: string[];
}

/**
 * Share to Facebook
 */
export function shareToFacebook(options: ShareOptions): void {
  const facebookShareUrl = `https://www.facebook.com/sharer/sharer.php?u=${encodeURIComponent(options.url)}`;
  window.open(facebookShareUrl, 'facebook-share', 'width=600,height=400');
}

/**
 * Share to Twitter
 */
export function shareToTwitter(options: ShareOptions): void {
  const hashtags = options.hashtags?.join(',') || '';
  const text = options.text || options.title;
  const twitterShareUrl = `https://twitter.com/intent/tweet?url=${encodeURIComponent(
    options.url
  )}&text=${encodeURIComponent(text)}${hashtags ? `&hashtags=${hashtags}` : ''}`;
  window.open(twitterShareUrl, 'twitter-share', 'width=600,height=400');
}

/**
 * Share to WhatsApp
 */
export function shareToWhatsApp(options: ShareOptions): void {
  const text = `${options.title}\n${options.url}`;
  const whatsappShareUrl = `https://wa.me/?text=${encodeURIComponent(text)}`;
  window.open(whatsappShareUrl, 'whatsapp-share', 'width=600,height=400');
}

/**
 * Share to LinkedIn
 */
export function shareToLinkedIn(options: ShareOptions): void {
  const linkedinShareUrl = `https://www.linkedin.com/sharing/share-offsite/?url=${encodeURIComponent(options.url)}`;
  window.open(linkedinShareUrl, 'linkedin-share', 'width=600,height=400');
}

/**
 * Share via Email
 */
export function shareViaEmail(options: ShareOptions): void {
  const subject = encodeURIComponent(options.title);
  const body = encodeURIComponent(`Check this out: ${options.text || options.title}\n\n${options.url}`);
  const emailShareUrl = `mailto:?subject=${subject}&body=${body}`;
  window.location.href = emailShareUrl;
}

/**
 * Copy direct link to clipboard
 */
export async function copyDirectLink(url: string): Promise<boolean> {
  try {
    if (navigator.clipboard && window.isSecureContext) {
      // Use modern Clipboard API if available
      await navigator.clipboard.writeText(url);
      return true;
    } else {
      // Fallback for older browsers
      const textArea = document.createElement('textarea');
      textArea.value = url;
      textArea.style.position = 'fixed';
      textArea.style.left = '-999999px';
      textArea.style.top = '-999999px';
      document.body.appendChild(textArea);
      textArea.focus();
      textArea.select();
      const success = document.execCommand('copy');
      document.body.removeChild(textArea);
      return success;
    }
  } catch (error) {
    console.error('Failed to copy link:', error);
    return false;
  }
}

/**
 * Open Web Share API if available (for mobile devices)
 */
export async function useWebShareAPI(options: ShareOptions): Promise<boolean> {
  if (!navigator.share) {
    return false;
  }

  try {
    await navigator.share({
      title: options.title,
      text: options.text,
      url: options.url,
    });
    return true;
  } catch (error) {
    if (error instanceof Error && error.name === 'AbortError') {
      // User cancelled the share
      return false;
    }
    console.error('Error using Web Share API:', error);
    return false;
  }
}

/**
 * Get the share URL for a post
 */
export function getPostShareUrl(postId: number | string, postSlug?: string): string {
  const baseUrl = typeof window !== 'undefined' ? window.location.origin : 'https://centimet2.com';

  if (postSlug) {
    return `${baseUrl}/posts/${postSlug}`;
  }

  return `${baseUrl}/posts/${postId}`;
}

/**
 * Handle social media share with appropriate platform
 */
export async function handleSocialShare(
  platform: string,
  postId: number,
  postTitle: string,
  postSlug?: string,
  postText?: string
): Promise<void> {
  const shareUrl = getPostShareUrl(postId, postSlug);

  const options: ShareOptions = {
    url: shareUrl,
    title: postTitle,
    text: postText || postTitle,
    hashtags: ['centimet2', 'marketplace'],
  };

  switch (platform.toLowerCase()) {
    case 'facebook':
      shareToFacebook(options);
      break;
    case 'twitter':
      shareToTwitter(options);
      break;
    case 'whatsapp':
      shareToWhatsApp(options);
      break;
    case 'linkedin':
      shareToLinkedIn(options);
      break;
    case 'email':
      shareViaEmail(options);
      break;
    case 'direct':
      const copied = await copyDirectLink(shareUrl);
      if (copied) {
        // Show a success toast/notification
        console.log('Link copied to clipboard!');
      } else {
        console.error('Failed to copy link');
      }
      break;
    default:
      console.warn(`Unknown share platform: ${platform}`);
  }
}

File: flutter/lib/src/utils/share_utils.dart
/// Share Utilities for Flutter
/// Handles sharing posts to various social media platforms

import 'package:share_plus/share_plus.dart';
import 'package:url_launcher/url_launcher.dart';

class ShareUtils {
  /// Get the base URL for the app
  static String get _baseUrl {
    return 'https://centimet2.com';
  }

  /// Get the post share URL
  static String getPostShareUrl({
    required int postId,
    String? postSlug,
  }) {
    if (postSlug != null && postSlug.isNotEmpty) {
      return '$_baseUrl/posts/$postSlug';
    }
    return '$_baseUrl/posts/$postId';
  }

  /// Share to Facebook
  static Future<void> shareToFacebook({
    required String postUrl,
    String? title,
  }) async {
    final facebookUrl = 'https://www.facebook.com/sharer/sharer.php?u=${Uri.encodeComponent(postUrl)}';
    if (await canLaunchUrl(Uri.parse(facebookUrl))) {
      await launchUrl(Uri.parse(facebookUrl), mode: LaunchMode.externalApplication);
    } else {
      throw 'Could not launch Facebook share';
    }
  }

  /// Share to Twitter
  static Future<void> shareToTwitter({
    required String postUrl,
    required String title,
    List<String>? hashtags,
  }) async {
    final text = Uri.encodeComponent(title);
    final hashtagString = hashtags != null && hashtags.isNotEmpty ? '&hashtags=${hashtags.join(',')}' : '';
    final twitterUrl = 'https://twitter.com/intent/tweet?url=${Uri.encodeComponent(postUrl)}&text=$text$hashtagString';

    if (await canLaunchUrl(Uri.parse(twitterUrl))) {
      await launchUrl(Uri.parse(twitterUrl), mode: LaunchMode.externalApplication);
    } else {
      throw 'Could not launch Twitter share';
    }
  }

  /// Share to WhatsApp
  static Future<void> shareToWhatsApp({
    required String postUrl,
    required String title,
  }) async {
    final message = Uri.encodeComponent('$title\n\n$postUrl');
    final whatsappUrl = 'https://wa.me/?text=$message';

    if (await canLaunchUrl(Uri.parse(whatsappUrl))) {
      await launchUrl(Uri.parse(whatsappUrl), mode: LaunchMode.externalApplication);
    } else {
      throw 'Could not launch WhatsApp share';
    }
  }

  /// Share to LinkedIn
  static Future<void> shareToLinkedIn({
    required String postUrl,
    String? title,
  }) async {
    final linkedinUrl = 'https://www.linkedin.com/sharing/share-offsite/?url=${Uri.encodeComponent(postUrl)}';

    if (await canLaunchUrl(Uri.parse(linkedinUrl))) {
      await launchUrl(Uri.parse(linkedinUrl), mode: LaunchMode.externalApplication);
    } else {
      throw 'Could not launch LinkedIn share';
    }
  }

  /// Share via Email
  static Future<void> shareViaEmail({
    required String postUrl,
    required String title,
    String? body,
  }) async {
    final subject = Uri.encodeComponent(title);
    final emailBody = Uri.encodeComponent(body ?? 'Check this out: $title\n\n$postUrl');
    final emailUrl = 'mailto:?subject=$subject&body=$emailBody';

    if (await canLaunchUrl(Uri.parse(emailUrl))) {
      await launchUrl(Uri.parse(emailUrl));
    } else {
      throw 'Could not launch email share';
    }
  }

  /// Copy link to clipboard and show native share sheet
  static Future<void> copyAndShare({
    required String postUrl,
    required String title,
    String? text,
  }) async {
    await SharePlus.instance.share(
      ShareParams(
        text: '${text ?? title}\n\n$postUrl',
        subject: title,
      ),
    );
  }

  /// Handle social share with specified platform
  static Future<void> handleShare({
    required String platform,
    required int postId,
    required String postTitle,
    String? postSlug,
    String? postText,
    List<String>? hashtags,
  }) async {
    final postUrl = getPostShareUrl(postId: postId, postSlug: postSlug);
    final shareText = postText ?? postTitle;

    switch (platform.toLowerCase()) {
      case 'facebook':
        await shareToFacebook(postUrl: postUrl, title: postTitle);
        break;
      case 'twitter':
        await shareToTwitter(
          postUrl: postUrl,
          title: postTitle,
          hashtags: hashtags ?? ['centimet2', 'marketplace'],
        );
        break;
      case 'whatsapp':
        await shareToWhatsApp(postUrl: postUrl, title: shareText);
        break;
      case 'linkedin':
        await shareToLinkedIn(postUrl: postUrl, title: postTitle);
        break;
      case 'email':
        await shareViaEmail(postUrl: postUrl, title: postTitle, body: shareText);
        break;
      case 'direct':
        await copyAndShare(
          postUrl: postUrl,
          title: postTitle,
          text: shareText,
        );
        break;
      default:
        throw 'Unknown share platform: $platform';
    }
  }

  /// Show share platform selection dialog
  static Future<String?> showShareDialog() async {
    return Future.value(null); // This will be implemented in the widget
  }
}
